File: code/packages/ruby/codegen-core/test/test_codegen_core.rb
# frozen_string_literal: true

require "minitest/autorun"
require "coding_adventures_codegen_core"
require "coding_adventures_interpreter_ir"

class CodegenCoreTest < Minitest::Test
  IR = CodingAdventures::InterpreterIr

  def test_default_registry_emits_all_requested_targets
    mod = IR::IIRModule.new(
      name: "demo",
      language: "test",
      functions: [
        IR::IIRFunction.new(
          name: "main",
          instructions: [
            IR::IIRInstr.new("const", "x", [1], "u8"),
            IR::IIRInstr.new("ret", nil, ["x"], "u8")
          ]
        )
      ]
    )
    registry = CodingAdventures::CodegenCore::BackendRegistry.default

    %i[pure_vm jvm clr wasm].each do |target|
      artifact = registry.compile(mod, target: target)
      assert_equal target, artifact.target
      assert_includes artifact.body, ".function main"
    end
  end
end

File: code/packages/lua/codegen_core/tests/test_codegen_core.lua
package.path = "../../interpreter_ir/src/?.lua;" .. "../../interpreter_ir/src/?/init.lua;" ..
    "../src/?.lua;" .. "../src/?/init.lua;" .. package.path

local ir = require("coding_adventures.interpreter_ir")
local codegen = require("coding_adventures.codegen_core")

describe("codegen_core", function()
    it("emits LANG text for registered targets", function()
        local fn = ir.IirFunction.new({
            name = "main",
            instructions = { ir.IirInstr.of("ret_void") },
            return_type = ir.Types.Void,
        })
        local mod = ir.IirModule.new({ name = "demo", functions = { fn }, entry_point = "main", language = "demo" })
        local artifact = codegen.BackendRegistry.default():compile(mod, "wasm")
        assert.equals("wasm", artifact.target)
        assert.is_truthy(artifact.body:find("%.function main"))
    end)
end)
